src/databricks/labs/ucx/source_code/redash.py

```diff
@@ -1,24 +1,171 @@
+                continue
+            for query in self.get_queries_from_dashboard(dashboard):
+                self._revert_query(query)
+            self._ws.dashboards.update(dashboard.id, tags=self._get_original_tags(dashboard.tags))
+
+    def _list_dashboards(self, dashboard_id: str | None) -> list[Dashboard]:
+        try:
+            if dashboard_id is None:
+                return list(self._ws.dashboards.list())
+            return [self._ws.dashboards.get(dashboard_id)]
+        except DatabricksError as e:
+            logger.debug(f"Cannot list dashboards: {e}")
+            return []
+
+    def delete_backup_queries(self, prompts: Prompts):
+        if not prompts.confirm("Are you sure you want to delete all backup queries?"):
+            return
+        for query in self._ws.queries.list():
+            # not a backup query
+            if query.tags is None or self.BACKUP_TAG not in query.tags:
+                continue
+            if query.id is None:
+                continue
+            try:
+                self._ws.queries.delete(query.id)
+            except NotFound:
+                continue
+
+    def _fix_query(self, query: Query):
         assert query.id is not None
         assert query.query is not None
-        query.query = self._fixer.apply(query.query)
-        self._ws.queries.update(
-            query.id,
+        # query already migrated
+        if query.tags is not None and self.MIGRATED_TAG in query.tags:
+            return
+        # create the backup folder for queries if it does not exist
+        try:
+            self._ws.workspace.get_status(self._backup_path)
+        except ResourceDoesNotExist:
+            self._ws.workspace.mkdirs(self._backup_path)
+        backup_query = self._ws.queries.create(
             data_source_id=query.data_source_id,
             description=query.description,
-            name=query.name,
-            options=query.options,
+            name=str(query.name) + "_original",
+            options=query.options.as_dict() if query.options is not None else None,
+            parent=self._backup_path,
             query=query.query,
             run_as_role=query.run_as_role,
+            tags=[self.BACKUP_TAG],
         )
+        from_table = FromTable(self._index, self._get_session_state(query))
+        new_query = from_table.apply(query.query)
+        try:
+            self._ws.queries.update(
+                query.id,
+                query=new_query,
+                tags=self._get_migrated_tags(query.tags, [f'backup:{backup_query.id}']),
+            )
+        except DatabricksError:
+            logger.error(f"Cannot upgrade {query.name}")
+            return
+
+    @staticmethod
+    def _get_session_state(query: Query) -> CurrentSessionState:
+        session_state = CurrentSessionState()
+        if query.options is None:
+            return session_state
+        if query.options.catalog:
```

Comment:
will the options always have a catalog and schema defined?

Reply:
no, it could be None

Reply:
I meant: will the attribute `catalog` and `schema` always be present, to avoid an AtributeError

Reply:
ah - both attributes will always be returned by the SDK